// @dhis2/app-runtime/build/cjs/lib.js
/* eslint-disable */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var appServiceData = require('@dhis2/app-service-data');
var appServiceConfig = require('@dhis2/app-service-config');
var React = _interopDefault(require('react'));

const Provider = ({
  config,
  children
}) => /*#__PURE__*/React.createElement(appServiceConfig.ConfigProvider, {
  config: config
}, /*#__PURE__*/React.createElement(appServiceData.DataProvider, null, children));
Provider.displayName = 'DHIS2RuntimeProvider';

Object.defineProperty(exports, 'CustomDataProvider', {
    enumerable: true,
    get: function () {
        return appServiceData.CustomDataProvider;
    }
});
Object.defineProperty(exports, 'DataMutation', {
    enumerable: true,
    get: function () {
        return appServiceData.DataMutation;
    }
});
Object.defineProperty(exports, 'DataProvider', {
    enumerable: true,
    get: function () {
        return appServiceData.DataProvider;
    }
});
Object.defineProperty(exports, 'DataQuery', {
    enumerable: true,
    get: function () {
        return appServiceData.DataQuery;
    }
});
Object.defineProperty(exports, 'useDataEngine', {
    enumerable: true,
    get: function () {
        return appServiceData.useDataEngine;
    }
});
Object.defineProperty(exports, 'useDataMutation', {
    enumerable: true,
    get: function () {
        return appServiceData.useDataMutation;
    }
});
Object.defineProperty(exports, 'useDataQuery', {
    enumerable: true,
    get: function () {
        return appServiceData.useDataQuery;
    }
});
Object.defineProperty(exports, 'useConfig', {
    enumerable: true,
    get: function () {
        return appServiceConfig.useConfig;
    }
});
exports.Provider = Provider;
